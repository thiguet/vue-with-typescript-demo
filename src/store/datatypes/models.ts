export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product {
    name: string;
    measure: keyof typeof Measures;
    qtd: number;
    minQtd: number;
    img?: string;
}

export interface User {
    name: string;
    email: string;
}

export interface LoginSubmit {
    name: string;
    pass: string;
}

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
}
