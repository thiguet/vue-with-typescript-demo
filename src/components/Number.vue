<template>
    <input
        :id="id"
        type="number"
        :name="id"
        class="input"
        :value="value"
        @input="onInput"
    />
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component({
    name: 'Number',
})
export default class Number extends Vue {
    @Prop({ required: true }) readonly id!: string;

    @Prop({ required: false, default: '0' }) readonly min!: number;

    @Prop({ required: false }) readonly max!: number;

    @Prop({ required: true }) readonly value!: string;

    @Prop({ required: true }) readonly setValue!: Function;

    private onInput(event: { target: { value: string } }) {
        this.setValue(event.target.value);
    }
}
</script>

<style scoped></style>
