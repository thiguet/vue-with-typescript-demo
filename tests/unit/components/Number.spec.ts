import { shallowMount } from '@vue/test-utils';
import Number from '@/components/Number.vue';
import faker from 'faker';
import { NumberInputProps } from './models';

describe('Number.vue', () => {
    let props: NumberInputProps;

    const build = () => {
        const wrapper = shallowMount(Number, {
            propsData: { ...props },
        });

        const input = () => wrapper.find('.input');

        return {
            wrapper,
            input,
            inputEl: (): HTMLInputElement => input().element as HTMLInputElement,
        };
    };

    beforeEach(() => {
        props = {
            id: faker.random.alphaNumeric(),
            value: faker.random.word(),
            min: faker.random.number(),
            max: faker.random.number(),
            setValue: jest.fn(),
        };
    });

    it('renders component', () => {
        props.id = 'someFixedValidId';
        props.value = 'someFixedValidValue';
        props.min = 0;
        props.max = 100;

        const { wrapper } = build();

        expect(wrapper).toMatchSnapshot();
    });

    it('renders main child components', () => {
        const { input } = build();

        expect(input().exists()).toBe(true);
    });

    it('passes props to the component', () => {
        const { wrapper } = build();

        const { id, setValue, value, min, max } = wrapper.props();

        expect(id).toBe(props.id);
        expect(value).toBe(props.value);
        expect(setValue).toBe(props.setValue);
        expect(min).toBe(props.min);
        expect(max).toBe(props.max);
    });

    it("sets input's type property to element by props", () => {
        const { input } = build();

        expect(input().attributes().type).toBe('number');
    });

    it("sets input's input propety to setValue fn", async () => {
        const { wrapper, inputEl, input } = build();

        const { setValue } = wrapper.props();

        const value = `${faker.random.number()}`;

        inputEl().value = value;

        await input().trigger('input', { data: value } as InputEventInit);

        expect(setValue).toHaveBeenCalledWith(value);
    });
});
