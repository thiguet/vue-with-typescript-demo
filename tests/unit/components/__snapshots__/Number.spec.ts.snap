// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number.vue renders component 1`] = `<input id="someFixedValidId" type="number" name="someFixedValidId" class="input">`;
